### ISSUE TITLE
Dimmers not responding well

### ISSUE THREAD

**Issue by hcoppen** [ilcato/homebridge-fibaro-home-center#237]
It looks like the dimmers are not set properly in the last release. See image below. When I ask Siri to set them to 25% it's looks like that is done but after that it's set back to 80%. When I do in manually slow is works fine.

<img width="1019" alt="image" src="https://github.com/ilcato/homebridge-fibaro-home-center/assets/5754201/18b3ebaf-d6c3-4b51-a028-ca1fc449b90a">

**Comment by mkz212** [comment]
@hcoppen 

- It works well for me on two different HC (HC 3 Lite and HC 2 Lite). 
- How do you set the poller period in config?
- Is the problem related to one device or all dimmers?

**Comment by hcoppen** [comment]
It's the HC2 the poller is set to default and yes it seems de problem is with the most of the dimmers. But I have the idee that it sometimes does work. I didn't had this issue before I moved to de latest version. The lights don't even change to 25% en back to 80%. They just stay on 25% I think. So I think the command is not reaching the HC2

<img width="818" alt="image" src="https://github.com/ilcato/homebridge-fibaro-home-center/assets/5754201/55704e27-6e86-43de-9202-432fcd6d720c">

**Comment by mkz212** [comment]
@hcoppen 

It's hard for me to reproduce the error that occurs for you. I checked at home on various dimmers, both: manually controlling in Apple Home and by voice via Siri and it works without the slightest problem. I don't see a problem in the code either. 

As for your settings, I have similar, I do not only have the add room name to device name option - I mention it because the fixes in the last version mainly concerned this option.

**Comment by mkz212** [comment]
@hcoppen on 1.5.0 set brighness to some value e.g 50% and after that turn off this dimmer (turn off, not set value to 0) - is it working for you?

**Comment by hcoppen** [comment]
No that's right when I set it to 50% and I turn it of it stays on.

**Comment by mkz212** [comment]
I checked with poller off and set to 2 s. I checked by controlling manually and using Siri. In any case, it works for me without a problem. 

The only fact is that Siri sends two commands each time (turnOn / turnOff + value) and the manually control (in Apple Home) sends the turnOn / turnOff command only when you change the value from 0 to more than 0 or from more than 0 to 0. When you disable lamp by clicking on icon it will send command turnOff, but when you slide down in detail view it will send value 0. But still works for me without a problem.

**Comment by hcoppen** [comment]
What version of dimmers are you using maybe that's the problem

**Comment by mkz212** [comment]
I also have, among others, 3.5 and they work without a problem.

And do you have the latest versions in iPhone / HomePod etc.?

Can you try with other poller period settings? 0 (disabled) or 10 sec.?

**Comment by hcoppen** [comment]
commando's are send but they go to 100% and then back to 25% in HC2 see image
<img width="884" alt="image" src="https://github.com/ilcato/homebridge-fibaro-home-center/assets/5754201/2503de84-0425-4e24-b384-c4ab092fbe3d">

**Comment by hcoppen** [comment]
If I turn poller off it works when going to 100% but when I go to 25% after it go's off

**Comment by hcoppen** [comment]
Yes I think it's with polling. When I set it to 10s it works the first time. But the second time it doesn't. But after 10s it works again.

**Comment by mkz212** [comment]
I understand well that all commands are sent correctly, i.e. they go to the Fibaro HC? The problem is that a moment later it is sent as if a second command that reverses this value?

**Comment by hcoppen** [comment]
But I am still wondering why version 1.5.0 worked well with Siri. The only issue was the 50% and turning off. So there must be something else

**Comment by mkz212** [comment]
https://github.com/ilcato/homebridge-fibaro-home-center/pull/189

This has changed between 1.5.0 and 1.5.1. The old code had a mistake - so the light cannot be turn off after changing the value. (What did you also say you couldn't do).

**Comment by hcoppen** [comment]
But why are there going 2 ‘on’ commands in 1.6.0 and only 1 ‘on’ command in 1.5.0.

**Comment by mkz212** [comment]
> But why are there going 2 ‘on’ commands in 1.6.0 / 1.51 and only 1 ‘on’ command in 1.5.0.

- I don't think it's a problem at all. For me, it is also the case that when I control using Siri, it sends two commands and when manually in Apple Home, it is one and so in both cases it works well.

- In your case, the commands are sent, right? And they reach HC? Only, moments later, the value is kind of reversed? Is the problem only about Siri? Does turning on and off also have this problem, or problem is only when setting the value?

- Does setting the poller period to 0 (disabled) change something, helps?

**Comment by hcoppen** [comment]
I will get back to your question tomorrow. I found out that turn on and off by hand doesn’t work well every time. Sometimes it jumps to 100% but the lights are 25% then after 2 seconds it shows 25%. Other times it jumps to 100% and stays at 100% and also the lights burn 100%. And other times it works as it should work.

I will get back to you tomorrow. Thanks for now

**Comment by mkz212** [comment]
@hcoppen 

I did a PR that should:
- solve the problem of blinking dimmers when setting values
- solve the problem of double commands when controlling Siri

If the author of the plugin merge this PR and releases a new version (probably at the beginning beta) then please test.

**Comment by hcoppen** [comment]
Okay I will test as soon as PR is approved and beta is available. Thanks for now!

**Comment by ilcato** [comment]
@hcoppen published 1.6.1-beta.0
@mkz212 see my comments on PR

**Comment by hcoppen** [comment]
I will test it this evening and give you feedback. Thanks

**Comment by hcoppen** [comment]
Hi changing the level works now with app and Siri. But I can't turn off the light anymore. Not by app and not by Siri. You can only turn it off by moving the level to 0. Als turning on the light will always move the level to 100%

**Comment by mkz212** [comment]
@hcoppen Thank you. Indeed it is. I did PR to fix it. The next beta should be ok.

@ilcato Please merge and release next beta.

**Comment by hcoppen** [comment]
Ok I will test again

**Comment by mkz212** [comment]
@hcoppen 

Thank you. 

When you will test it, please pay attention:
- when setting the light brightness value (moving the slider) one command or several are sent? The light smoothly passes e.g. from 20% to 80% or, for example, it is set to 35% then 47%, etc. (it blinks) and only at the end it is 80%?
- similarly with roller / blinds - moving the slider, for example from 20% to 80%, one command or several are sent? The roller / blinds smoothly changes positions or stops along the way?

**Comment by hcoppen** [comment]
With the current beta version the command is only sent when I release the slider after moving. So I go from 30% to 60% slowly it's only send when I release the slider at 60% 'Set 60%'. After 1 or 2 seconds it sends 'on' and then '60%' that's what your log says. 

<img width="598" alt="image" src="https://github.com/ilcato/homebridge-fibaro-home-center/assets/5754201/3262bad3-f774-4e5c-a4aa-f9aced0c8608">

**Comment by ilcato** [comment]
> @ilcato Please merge and release next beta.

@mkz212, the new PR only changes the variable name.

**Comment by mkz212** [comment]
@ilcato No. It add clearTimeout.

**Comment by ilcato** [comment]
@hcoppen @mkz212 published beta.1

**Comment by mkz212** [comment]
For me now it is working as expected. 

There is a slight delay in setting the value but there is no blinking (in the case of dimmers) and jamming (in the case of blinds).

Unfortunately, Apple Home works in such a way that it sends commands when you move the slider, not only when you raise your finger.

**Comment by hcoppen** [comment]
It's is turning off now but when I turn it off and back on it always go's to 100%. So turning is off while it's 30% will go to 100% when I turn it back on

**Comment by mkz212** [comment]
@hcoppen In version 1.6.0 it also happens like this.

**Comment by mkz212** [comment]
@ilcato So far, I have come to the conclusion that when the dimmer is turned off, the value is set to 0%. And since it's 0%, it's like there's no value so it turns on 100%.

**Comment by hcoppen** [comment]
@ilcato  Okay, but is it not an option to use only 'turnOff' or 'turnOn' option on Fibaro when turning on or off and the 'setValue' option when moving the slider in homekit. POST /api/devices/{diviceID}/action/turnOn and turnOff should work. So by using this command it will remember its state. When using the slider in homeKit and the value is > 0 and < 99/100 u should use the setValue command.

**Comment by mkz212** [comment]
> @ilcato Okay, but is it not an option to use only 'turnOff' or 'turnOn' option on Fibaro when turning on or off and the 'setValue' option when moving the slider in homekit. POST /api/devices/{diviceID}/action/turnOn and turnOff should work. So by using this command it will remember its state. When using the slider in homeKit and the value is > 0 and < 99/100 u should use the setValue command.

This is not the point at all. The commands in HomeKit and in this plugin are correct.

The problem is that Apple stores two values: On (true/false) and Value (0-100%). If the Value is set to 0, it will set 100% at startup. Fibaro, on the other hand, complicated the case and Value uses only when the dimmer is on. If the Value is 0, it means that it is turned off. And the last value is stored elsewhere.

Example, when we turn of dimmer with value 73%:
- in HomeKit, On will be false, Value will be 73%. 
- in Fibaro, Value will be 0. And somewhere else they store last used value to use when you turn on dimmer.

**Comment by mkz212** [comment]
@hcoppen 

You can check. Get API response for device:
Open in browser: http://fibaro-ip/api/devices/DEVICE-ID (replace FIBARO-IP with your Home Center IP and DEVICE-ID with device ID) and login.

Next turn on / off dimmer. You will see that it does not matter if you turn off through Homekit or Fibaro, the value will be 0.

**Comment by mkz212** [comment]
I have HC 3 Lite and the latest firmware 5.150.18 and it happens like this. I will check on HC2 Lite. And you @hcoppen what HC and firmware do you have?

**Comment by hcoppen** [comment]
> > @ilcato Okay, but is it not an option to use only 'turnOff' or 'turnOn' option on Fibaro when turning on or off and the 'setValue' option when moving the slider in homekit. POST /api/devices/{diviceID}/action/turnOn and turnOff should work. So by using this command it will remember its state. When using the slider in homeKit and the value is > 0 and < 99/100 u should use the setValue command.
> 
> This is not the point at all. The commands in HomeKit and in this plugin are correct.
> 
> The problem is that Apple stores two values: On (true/false) and Value (0-100%). If the Value is set to 0, it will set 100% at startup. Fibaro, on the other hand, complicated the case and Value uses only when the dimmer is on. If the Value is 0, it means that it is turned off. And the last value is stored elsewhere.
> 
> Example, when we turn of dimmer with value 73%:
> 
> * in HomeKit, On will be false, Value will be 73%.
> * in Fibaro, Value will be 0. And somewhere else they store last used value to use when you turn on dimmer.

Okay, clear. So after turning off the poller takes the value from Fibaro and that will be 0. Is it possible to know if only turnoff or turnon button is clicked in homekit? If so you could only use the /action/turnOn command then read the level from Fibaro and set it in HomeKit.

**Comment by hcoppen** [comment]
> I have HC 3 Lite and the latest firmware 5.150.18 and it happens like this. I will check on HC2 Lite. And you @hcoppen what HC and firmware do you have?

latest 4.6.30

**Comment by mkz212** [comment]
@hcoppen 

- We still don't understand each other.
- Poller has nothing to do with it (it happens wit on and off). 
- The command also does not. But to make you know, the commands are currently being distributed well - recognise on, off, set value. 
- The problem is that at the start the value is always 0, and when you want to turn on with the value 0, it is logical that you do not turn on the value 0 so it set 100.
- The solution is such that you have to find where Fibaro stores the value for the DISABLED (turned off) dimmer and use this value as the dimmer is turned off.

**Comment by hcoppen** [comment]
> @hcoppen
> 
> We still don't understand each other:
> 
> * Poller has nothing to do with it (it happens with turned on and off).
> * The command also does not. But to make you know, the commands are currently being distributed well - recognise on, off, set value.
> * The problem is that at the start the value is always 0, and when you want to turn on with the value 0, it is logical that you do not turn on the value 0 so it set 100.
> * The solution is such that you have to find where Fibaro stores the value for the DISABLED (turned off) dimmer and use this value as the dimmer is turned off.

regarding to your second point are you turning on the device by setting the value (POST  /api/devices/{diviceID}/action/setValue)? If so then it's clear you can't set 0 for turning on. Or do you use the POST /api/devices/{diviceID}/action/turnOn call?

**Comment by mkz212** [comment]
> > @hcoppen
> > We still don't understand each other:
> > 
> > * Poller has nothing to do with it (it happens with turned on and off).
> > * The command also does not. But to make you know, the commands are currently being distributed well - recognise on, off, set value.
> > * The problem is that at the start the value is always 0, and when you want to turn on with the value 0, it is logical that you do not turn on the value 0 so it set 100.
> > * The solution is such that you have to find where Fibaro stores the value for the DISABLED (turned off) dimmer and use this value as the dimmer is turned off.
> 
> regarding to your second point are you turning on the device by setting the value (POST /api/devices/{diviceID}/action/setValue)? If so then it's clear you can't set 0 for turning on. Or do you use the POST /api/devices/{diviceID}/action/turnOn call?

It doesn't matter. If you make a setValue with a value of 0 it will be 100. If you make a turnOn with a value of 0, the effect will be the same. The problem is NOT in setting the value but in reading it!

Before this plugin set value it will first read value. For turned off dimmer it will always read Value = 0. This is because Fibaro in API, for turned off (only for turned off) dimmers, store last used value somewhere else - not in Value field. It use field Value only for enabled dimmers.

**Comment by hcoppen** [comment]
> action/turnOn

But that's what I mean you have to POST /api/devices/{diviceID}/action/turnOn  with empty params {"args":[]} This will restore the last value of the dimmer in Fibaro

**Comment by mkz212** [comment]
@hcoppen 

The setValue command must be sent, otherwise when you change the brightness with the slider in Apple Home, the new value would not go to Fibaro. I don't know if it can be distinguished somehow, but it's certainly easier and more reasonable to just read the correct value from the API.

**Comment by hcoppen** [comment]
> @hcoppen
> 
> The setValue command must be sent, otherwise when you change the brightness with the slider in Apple Home, the new value would not go to Fibaro. I don't know if it can be distinguished somehow, but it's certainly easier and more reasonable to just read the correct value from the API.

Yes, that's why I asked if you can distinguish between switching on (click on button) and the slider. When the slider is used you can use the setValue if you only click on button, to turn dimmer on, you could use the turnOn command

**Comment by mkz212** [comment]
> > @hcoppen
> > The setValue command must be sent, otherwise when you change the brightness with the slider in Apple Home, the new value would not go to Fibaro. I don't know if it can be distinguished somehow, but it's certainly easier and more reasonable to just read the correct value from the API.
> 
> Yes, that's why I asked if you can distinguish between switching on (click on button) and the slider. When the slider is used you can use the setValue if you only click on button, to turn dimmer on, you could use the turnOn command

I don't know if it possible. And even if it is:
- different way Apple Home sends command when you click on icon, different when you use slider, and different when use Siri.
- in a situation when you change the value in the Fibaro application or using the button, the Home application will not retrieve the new value correctly.

**Comment by mkz212** [comment]
@hcoppen If you would like to help, find where in the API, Fibaro stores the value but for the DISABLED Dimmer.

**Comment by ilcato** [comment]
@mkz212, @hcoppen,  I didn't follow all the discussion but I think the problem is in the poller mechanism: when the dimmer is turned off the poller API update both the status of the On and Brightness characteristics --> Off and 0%.
The solution for me is to ignore the brightness update when the Off update is received (I'm assuming the Off update comes first). Let me check ...

**Comment by mkz212** [comment]
> @mkz212, @hcoppen, I didn't follow all the discussion but I think the problem is in the poller mechanism: when the dimmer is turned off the poller API update both the status of the On and Brightness characteristics --> Off and 0%. The solution for me is to ignore the brightness update when the Off update is received (I'm assuming the Off update comes first). Let me check ...

with poller disabled it also get wrong value

**Comment by ilcato** [comment]
> > @mkz212, @hcoppen, I didn't follow all the discussion but I think the problem is in the poller mechanism: when the dimmer is turned off the poller API update both the status of the On and Brightness characteristics --> Off and 0%. The solution for me is to ignore the brightness update when the Off update is received (I'm assuming the Off update comes first). Let me check ...
> 
> with poller disabled it also get wrong value

For some reason, disabling poller is not working.

**Comment by mkz212** [comment]
> > > @mkz212, @hcoppen, I didn't follow all the discussion but I think the problem is in the poller mechanism: when the dimmer is turned off the poller API update both the status of the On and Brightness characteristics --> Off and 0%. The solution for me is to ignore the brightness update when the Off update is received (I'm assuming the Off update comes first). Let me check ...
> > 
> > 
> > with poller disabled it also get wrong value
> 
> For some reason, disabling poller is not working.

- Maybe every time before plugin send set command it first get value?
- I believe that best option is in getFunctions and in poller do something like: if dimmer is on get value from Value, else get from other field (that where fibaro store last used value)

**Comment by ilcato** [comment]
Try beta.2, it seems to work

**Comment by mkz212** [comment]
Beta 2 with disabled poller works.

**Comment by ilcato** [comment]
I have poller enabled

**Comment by mkz212** [comment]
Ok with enabled and disabled poller it works great!! 

I don't sleep here half a night and I think how to solve it ... the boss comes and probably one hand, drinking coffee, wrote the code and works 😉👍

**Comment by ilcato** [comment]
@mkz212, I opened an issue on the poller disabling mechanism not working. Probably it is something relating to the way you implemented the check on theHome Center availability.
I will also reduce the timeout for the brightness from 500 to 300ms.

**Comment by hcoppen** [comment]
Great work guys It also works for me! Just tested it!

**Comment by mkz212** [comment]
@ilcato 

- I did PR - should fix disabling poller. 
- I tested with 300 ms for dimmers and for me it was to less - multiple values was send - so maybe 400 ms?
- You can test less value for roller shutters / blind, now it is 1500 ms.

**Comment by mkz212** [comment]
> Great work guys It also works for me! Just tested it!

@hcoppen 
Thanks for reporting bugs and for suggestions and for testing! Great job! 👍

**Comment by ilcato** [comment]
> * I tested with 300 ms for dimmers and for me it was to less - multiple values was send - so maybe 400 ms?

@mkz212 is it still necessary the whole mechanism ? Isn't the initial issue caused by the poller? What about instead of ignoring multiple setValues in sequence blocking the pollerUpdate and reenabling it after the last change ? You can reuse the same timeout structure.

**Comment by mkz212** [comment]
> > * I tested with 300 ms for dimmers and for me it was to less - multiple values was send - so maybe 400 ms?
> 
> @mkz212 is it still necessary the whole mechanism ? Isn't the initial issue caused by the poller? What about instead of ignoring multiple setValues in sequence blocking the pollerUpdate and reenabling it after the last change ? You can reuse the same timeout structure.

I think it's not poller's fault. It's just that Apple sends too many commands when moving the brightness slider - it should only one at the end of the move, once you raise your finger, but it is not.

**Comment by ilcato** [comment]
I think that it is correct to send update for each slider movement, after all this is the behaviour of a physical slider. And the Fibaro API is able to cope with multiple update. The problem is in the asynchronous and delayed feedback coming from the poller. If we disable poller updates while a slider is moving I think the problem disappear.

**Comment by mkz212** [comment]
> I think that it is correct to send update for each slider movement, after all this is the behaviour of a physical slider. And the Fibaro API is able to cope with multiple update. The problem is in the asynchronous and delayed feedback coming from the poller. If we disable poller updates while a slider is moving I think the problem disappear.

We need to test.

**Comment by mkz212** [comment]
@ilcato 

- I tested and when I control directly through Fibaro it sends only one command - when I let go of the slider.
- Sending several commands (when using Home slider) causes a blink.
- I don't think that sending a few commands is ok.

That's why I would leave it as it is now. If Apple changes the action - so that the command is sent after lifting the finger - then we can eliminate timeout.

**Comment by mkz212** [comment]
@ilcato 

I personally definitely prefer to have moments of delay, but so that it does not blink and jam. And now when you fixed that it's not always set to 100%, it's already great!

**Comment by hcoppen** [comment]
I think it works great now guys! Even the RGBW controller works great now. Thanks for your quick support!

**Comment by ilcato** [comment]
> @ilcato
> 
> I personally definitely prefer to have moments of delay, but so that it does not blink and jam. And now when you fixed that it's not always set to 100%, it's already great!

The logic I proposed was already present in method `setCharacteristicValue` of `FibaroAccessory` class. I slightly modified it. Please test the new beta. Will close this issue after your feedback.

**Comment by mkz212** [comment]
@ilcato 1.7.0-beta.0 seems to be ok for me! 👍 I did PR to quiet warnings.

**Comment by hcoppen** [comment]
@ilcato 1.7.0-beta.0 works for me too! 👍

**Comment by patrikz** [comment]
Hello,
I see that this issue is closed. However I found this when I was searching for symptoms where there is issues in dimming light in Apple Home and fibaro. It has been working fine for me but I have encountered problems with dimming, which has worked fine before. There are long delays both reading and setting new dimmer values using Apple Home App. Sending commands with Siri seem to work fine.

I use Fibaro HC3 running v 5.170.16
homebridge Figaro Home Center v 3.2.0.

As I understand this thread was this corrected with an update in this plug-in. But did you also have to change anything in the configuration to make it work properly?
